include/fmt/core.h

```diff
@@ -2204,20 +2204,32 @@ constexpr auto to_ascii(Char c) -> char {
   return c <= 0xff ? static_cast<char>(c) : '\0';
 }
 
+// Returns the length of a codepoint. Returns 0 for invalid codepoints.
 FMT_CONSTEXPR inline auto code_point_length_impl(char c) -> int {
   return "\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\1\0\0\0\0\0\0\0\0\2\2\2\2\3\3\4"
       [static_cast<unsigned char>(c) >> 3];
 }
+// Returns the length of a codepoint. Returns 1 for invalid codepoints.
+// This is equivalent to
+//
+// int len = code_point_length_impl(c);
+// return len + !len;
+//
+// This is useful because it allows the compiler to check that the
+// length is within the range [1, 4]
+FMT_CONSTEXPR inline auto code_point_length_impl_2(char c) -> int {
+  return static_cast<int>((0x3a55000000000000ull >> (2 * (static_cast<unsigned char>(c) >> 3))) & 0x3) + 1;
+}
```

Comment:
I don't think we need both `code_point_length_impl` and `code_point_length_impl_2`, let's merge this logic into `code_point_length_impl` and make the latter take `unsigned char` to have one case instead of 3.

Reply:
On a second thought, this logic can be moved directly into `code_point_length` and both impl functions removed.

Reply:
The return value of ``code_point_length`` is used as ``len`` and as ``len + !len``.

First case (first and last lines):
https://github.com/fmtlib/fmt/blob/3daf33837c98f95a7763459e06ce443abb883726/include/fmt/format.h#L654-L664

Second case:
https://github.com/fmtlib/fmt/blob/3daf33837c98f95a7763459e06ce443abb883726/include/fmt/core.h#L2215-L2223

Reply:
Ah, right. Then I think we should merge `code_point_length_impl` into `utf8_decode` and `code_point_length_impl_2` into `code_point_length`. Unless, of course, there is a similar false positive in `utf8_decode`.